internal/debug/flags.go

```diff
@@ -151,25 +173,62 @@ func Setup(ctx *cli.Context) error {
 		ostream = log.StreamHandler(output, log.TerminalFormat(usecolor))
 	}
 	glogger.SetHandler(ostream)
+
 	// logging
-	log.PrintOrigins(ctx.GlobalBool(debugFlag.Name))
-	glogger.Verbosity(log.Lvl(ctx.GlobalInt(verbosityFlag.Name)))
-	glogger.Vmodule(ctx.GlobalString(vmoduleFlag.Name))
-	glogger.BacktraceAt(ctx.GlobalString(backtraceAtFlag.Name))
+	if ctx.GlobalIsSet(legacyDebugFlag.Name) {
+		log.PrintOrigins(ctx.GlobalBool(legacyDebugFlag.Name))
+		log.Warn("The flag --debug is deprecated and will be removed in the future, please use --log.debug")
+	}
+	if ctx.GlobalIsSet(debugFlag.Name) {
+		log.PrintOrigins(ctx.GlobalBool(debugFlag.Name))
+	}
```

Comment:
There configs are incorrect. `GlobalIsSet` will only return `true` if the user explicitly set it to something, but they will remain `false` if the defaults are left as is. This means that all the flag initialization will be skipped by the code in this PR.

Reply:
Given that all these calls have side effects, it's a bit messy, you have 4 cases that you need to handle

- no flag is set
- both are set
- legacy is set
- new flag is set

You handled 2 ok-ish. The one where nothing is set is a noop now, and the one were both is set may work, but it might have weird side effects.

What you could do is:

```go
debug := ctx.GlobalBool(debugFlag.Name)
if ctx.GlobalIsSet(legacyDebugFlag.Name) {
    log.Warn("The flag --debug is deprecated and will be removed in the future, please use --log.debug")
    debug = ctx.GlobalBool(legacyDebugFlag.Name)
}
if ctx.GlobalIsSet(debugFlag.Name) {
    debug := ctx.GlobalBool(debugFlag.Name)
}
log.PrintOrigins(debug)
```

Reply:
Same for all the flags below

Reply:
Or

```go
debug := ctx.GlobalBool(debugFlag.Name)
if ctx.GlobalIsSet(legacyDebugFlag.Name) && !ctx.GlobalIsSet(debugFlag.Name) {
    log.Warn("The flag --debug is deprecated and will be removed in the future, please use --log.debug")
    debug = ctx.GlobalBool(legacyDebugFlag.Name)
}
log.PrintOrigins(debug)
```

Though this will not warn you if both are set, so the former might be more correct, even if a bit unwieldy.

Reply:
I had applied the solution you suggested to `memprofilerate` and `blockprofilerate` further down. But something I might be missing is whether the initialization methods should be invoked even when the value is empty? i.e. `debug` is by default false, does `log.PrintOrigins` has to be called explicitly even in that case? Because the only flag that has a non-empty default is `verbosity` which is initialised in `init()`.

Reply:
Aha now I see that `glogger.Vmodule` and `glogger.BacktraceAt` set some other variables that might have side-effects if unset. Will fix.